pkg/clustermesh/types/addressing.go

```diff
@@ -96,3 +111,90 @@ func MustParseAddrCluster(s string) AddrCluster {
 func (ac AddrCluster) Addr() netip.Addr {
 	return ac.addr
 }
+
+// Equal returns true when given AddrCluster has a same IP address and ClusterID
+func (ac0 AddrCluster) Equal(ac1 AddrCluster) bool {
+	return ac0.addr == ac1.addr && ac0.clusterID == ac1.clusterID
+}
+
+// Less compares ac0 and ac1 and returns true if ac0 is lesser than ac1
+func (ac0 AddrCluster) Less(ac1 AddrCluster) bool {
+	// First, compare the IP address part
+	if ret := ac0.addr.Compare(ac1.addr); ret == -1 {
+		return true
+	} else if ret == 1 {
+		return false
+	} else {
+		// If IP address is the same, compare ClusterID
+		return ac0.clusterID < ac1.clusterID
+	}
+}
+
+// This is an alias of Equal which only exists for satisfying deepequal-gen
+func (ac0 *AddrCluster) DeepEqual(ac1 *AddrCluster) bool {
+	return ac0.Equal(*ac1)
+}
+
+// DeepCopyInto copies in to out
+func (in *AddrCluster) DeepCopyInto(out *AddrCluster) {
+	if out == nil {
+		return
+	}
+	out.addr = in.addr
+	out.clusterID = in.clusterID
+}
+
+// DeepCopy returns a new copy of AddrCluster
+func (in *AddrCluster) DeepCopy() *AddrCluster {
+	out := new(AddrCluster)
+	in.DeepCopyInto(out)
+	return out
+}
+
+// String returns the string representation of the AddrCluster. If
+// AddrCluster.clusterID = 0, it returns bare IP address string. Otherwise, it
+// returns IP string + "@" + ClusterID (e.g. 10.0.0.1@1)
+func (ac AddrCluster) String() string {
+	if ac.clusterID == 0 {
+		return ac.addr.String()
+	}
+	return ac.addr.String() + "@" + strconv.FormatUint(uint64(ac.clusterID), 10)
+}
+
+// Is4 reports whether IP address part of AddrCluster is an IPv4 address.
+func (ac AddrCluster) Is4() bool {
+	return ac.addr.Is4()
```

Comment:
Not sure if this would be net benefit, but if `addr` was an unnamed member of `AddrCluster`, we would not need to define these "forwarder" functions, as all functions on `addr` would also work on `AddrCluster` type, if not explicitly overridden (like we want for `String()` etc.).

Reply:
I considered that but concluded that it is safer not to do embedding because if we do that, we implicitly "support" methods that are not intended to be used for `AddrCluster`. To avoid that, we need to override those functions explicitly. That's ok, but that means for every Go version update, we need to check if `netip.Addr` introduces the new method we need to override or not. Otherwise, someone may misuse those functions. Maybe I'm too worried?